File: python/second_minimum_node_in_binary_tree.py
from heapq import heappop, heappush

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def addNodesToSet(self, root: TreeNode, elements: set) -> None:
        if root is None: return
        elements.add(root.val)
        self.addNodesToSet(root.left, elements)
        self.addNodesToSet(root.right, elements)

    def unique_elements(self, root: TreeNode) -> set:
        elements = set()
        self.addNodesToSet(root, elements)
        return elements

    def findSecondMinimumValue(self, root: TreeNode) -> int:
        elements = self.unique_elements(root)
        if len(elements) < 2: return -1
        heap = []
        for element in elements:
            heappush(heap, element)
        heappop(heap)
        return heappop(heap)

File: src/SecondMinimumNodeInBinaryTree.java
import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Set;

public class SecondMinimumNodeInBinaryTree {
    public int findSecondMinimumValue(TreeNode root) {
        Set<Integer> elements = treeToSet(root);
        if (elements.size() < 2) return -1;
        Queue<Integer> queue = setToMinHeap(elements);
        queue.poll();
        return queue.peek();
    }

    private static Set<Integer> treeToSet(TreeNode root) {
        Set<Integer> result = new HashSet<>();
        addTreeNodesToSet(root, result);
        return result;
    }

    private static void addTreeNodesToSet(TreeNode root, Set<Integer> set) {
        if (root == null) return;
        set.add(root.val);
        addTreeNodesToSet(root.left, set);
        addTreeNodesToSet(root.right, set);
    }

    private static Queue<Integer> setToMinHeap(Set<Integer> elements) {
        Queue<Integer> queue = new PriorityQueue<>();
        for (int element : elements) {
            queue.add(element);
        }
        return queue;
    }
}
